core/dbt/context/providers.py

```diff
@@ -732,12 +733,28 @@ def invocation_args_dict(self):
         return args_to_dict(self.config.args)
 
     @contextproperty
-    def _sql_results(self) -> Dict[str, AttrDict]:
+    def _sql_results(self) -> Dict[str, Optional[AttrDict]]:
         return self.sql_results
 
     @contextmember
     def load_result(self, name: str) -> Optional[AttrDict]:
-        return self.sql_results.get(name)
+        if name in self.sql_results:
+            # handle the special case of "main" macro
+            if name == "main":
+                return self.sql_results["main"]
+
+            # handle a None, which indicates this name was populated but has since been loaded
+            elif self.sql_results[name] is None:
+                raise MacroResultAlreadyLoadedError(name)
+
+            # Handle the regular use case
+            else:
+                ret_val = self.sql_results[name]
+                self.sql_results[name] = None
+                return ret_val
+        else:
+            # Handle trying to load a result that was never stored
```

Comment:
@jtcohen6 Just pointing out that this, while it maintains parity to the previous code, doesn't seem like the best way to handle calls to `load_result` in instances where there is no matching key in `sql_results`.  Should we throw an error or something else here or have people possibly baked this behavior into their workflows and we should just let it slide?

Reply:
You raise a good point! After giving it a bit of thought, I do think we want this to keep returning `None`, rather than raising, because we do a first pass through this code **at parse time** — at which time, the `statement` has not been called, and so there's no matching value in `sql_results`.

Basically, this would start raising during parsing:
```sql
{% call statement('my_statement') %}
  select 1 as id
{% endcall %}
{% set res = load_result('my_statement') %}
```
We could insist that users stick the whole thing in an `if execute` block:
```
{% if execute %}
    {% call statement('my_statement') %}
      select 1 as id
    {% endcall %}
    {% set res = load_result('my_statement') %}
{% else %}
    {% set res = ... %}
{% endif %}
```
But that feels like putting more of the brunt on them — even if it would make it easier to catch instances where you just had a typo, e.g. `load_result('my_statment')`.

(Through a bit of cleverness, even if users try to operate on, or access values of the result while it's `None` - e.g. `res['table']` - at parse time, we catch and ignore those Jinja Undefined exceptions. I think.)

Reply:
Ok sounds good-- We'll leave it as is.